### ISSUE TITLE
add early stopping to backprop

### ISSUE THREAD

**Issue by lacava** [lacava/feat#152]
stop backprop when validation loss stops improving
